### ISSUE TITLE
Handle alerts and prompts

### ISSUE THREAD

**Issue by chrsmith** [chrsmith/selenium-test#27]
Original [issue 27](https://code.google.com/p/selenium/issues/detail?id=27) created by chrsmith on 2007-10-31T16:55:04.000Z:

Many sites use javascript alerts and prompts to communicate with the user.
It should be possible to test these using WebDriver.

**Comment by chrsmith** [comment]
Comment [#1](https://code.google.com/p/selenium/issues/detail?id=27#c1) originally posted by chrsmith on 2008-04-22T15:50:11.000Z:

Yes, this will be a great feature!!
Also have a look at watij!!

**Comment by chrsmith** [comment]
Comment [#2](https://code.google.com/p/selenium/issues/detail?id=27#c2) originally posted by chrsmith on 2008-06-04T15:06:49.000Z:

Yes it would be great!
I'm using WebDriver (r280) to test a website and I am stuck because of a javascript
"confirm" modal window. Is there any known workaround that could help me? I just want
to go through this window by automatically clicking "ok", I actually do not care
about what is displayed. Selenium is able to do that but I prefer WebDriver. Thanks
for your help!

**Comment by chrsmith** [comment]
Comment [#3](https://code.google.com/p/selenium/issues/detail?id=27#c3) originally posted by chrsmith on 2008-08-22T06:53:19.000Z:

Remark: login-dialog or download-dialog belongs to this sort of prompts too...
        (a workaround with sendKeys(Keys.ESCAPE) doesn't work)

**Comment by chrsmith** [comment]
Comment [#4](https://code.google.com/p/selenium/issues/detail?id=27#c4) originally posted by chrsmith on 2008-08-22T07:13:49.000Z:

A click on a <a href='...' target='_blank'> opens new windows/tabs...

A click on a <a href='javascript:openpopup();' onclick='return openpopup();'>
   opens new popups

It should be possible to prevent/control opening new windows/tabs/popup.

Possibilities:
 1. override target attribut
 2. callback for controlling new windows/tabs/popup
    (like IE: CHtmlView::OnNewWindow2, CHtmlView::OnBeforeNavigate2
     or CHtmlView::OnShowMessage)
 3. an excellent solution would be the possibility to redirect to my own window, 
    like the IE-code:
    void CMyHtmlView::OnNewWindow2(LPDISPATCH* ppDisp, BOOL* Cancel)
    {
       SetSilent(TRUE);
       *ppDisp = this->m_pBrowserApp;
    }

**Comment by chrsmith** [comment]
Comment [#5](https://code.google.com/p/selenium/issues/detail?id=27#c5) originally posted by chrsmith on 2008-08-22T09:23:18.000Z:

This bug is to do with handling alerts, prompts and "confirms" from javascript. The
aim is for this to appear as something like:

Alert alert = driver.switchTo().alert();
alert.accept();

Preventing new windows from being opened is outside the scope of this issue.

**Comment by chrsmith** [comment]
Comment [#6](https://code.google.com/p/selenium/issues/detail?id=27#c6) originally posted by chrsmith on 2008-12-18T19:38:52.000Z:

The following workaround made it for me:

((JavascriptExecutor) driver).executeScript("window.confirm = function(msg) { return
true; }");

**Comment by chrsmith** [comment]
Comment [#7](https://code.google.com/p/selenium/issues/detail?id=27#c7) originally posted by chrsmith on 2009-01-20T02:52:59.000Z:

Will this include support for modal dialogs?  I'm referring to the javascript
function showModalDialog (I think IE, FF and safari all support it now)

**Comment by chrsmith** [comment]
Comment [#8](https://code.google.com/p/selenium/issues/detail?id=27#c8) originally posted by chrsmith on 2009-02-09T15:32:54.000Z:

Do you have plans to implement this soon, because it has become an important issue
for us?  I need to know when an alert has been popped up and dismiss it.  It would be
nice to read the alert's text too.  I'm starting to think about checking out the
source and having a go myself.  Do you think it's a tough job?

**Comment by chrsmith** [comment]
Comment [#9](https://code.google.com/p/selenium/issues/detail?id=27#c9) originally posted by chrsmith on 2009-02-09T18:07:43.000Z:

I'd be delighted if someone else had a go at making this work :) As a short term fix,
you may consider overriding "window.alert" using the JavascriptExecutor.

**Comment by chrsmith** [comment]
Comment [#10](https://code.google.com/p/selenium/issues/detail?id=27#c10) originally posted by chrsmith on 2009-04-08T13:35:28.000Z:

I started automating a web application using webdriver.During execution of some
testcases i am getting java alerts.When i used
Alert alert = driver.switchTo().alert();
alert.accept();
I am getting error message The method alert() is undefined for the type WebDriver.
TargetLocator. Is the Issue no 27 is fixed?.Can you tell me what was the problem in
my code.


With Thanks & Regards,

Mayur Bongirwar

**Comment by chrsmith** [comment]
Comment [#11](https://code.google.com/p/selenium/issues/detail?id=27#c11) originally posted by chrsmith on 2009-04-20T15:26:21.000Z:

would be great if we could handle this without overrriding the window.confirm. After
all this is the point of webdriver, to test this stuff.

**Comment by chrsmith** [comment]
Comment [#12](https://code.google.com/p/selenium/issues/detail?id=27#c12) originally posted by chrsmith on 2009-04-24T19:53:44.000Z:

Would be great to help automate testing!

**Comment by chrsmith** [comment]
Comment [#13](https://code.google.com/p/selenium/issues/detail?id=27#c13) originally posted by chrsmith on 2009-04-25T13:03:16.000Z:

as a hack you could also inject some javascript to make an html POST or GET when the
js popup is shown. below is just my thoughts, haven't tried it yet

something like:

window.confirm=function(msg){
var winId = window._webdriver_handle_id;
var http = new XMlHttpRequest();
//set location to listening server
//send alert msg, win id
//return true/false
}

The window id is put in the window by webdriver. You'll have to write some  code to
do this. This id allows you to identify what window the alert came from, so you can
check the correct window called it. You then validate the msg. 

You'll also need to open up an http listener in your test code.

**Comment by chrsmith** [comment]
Comment [#14](https://code.google.com/p/selenium/issues/detail?id=27#c14) originally posted by chrsmith on 2009-05-05T17:19:09.000Z:

Has anyone found a workaround for this?  I've tried all the suggestions on this page,
but none of them get me past the javascript alert boxes.

Thanks!

**Comment by chrsmith** [comment]
Comment [#15](https://code.google.com/p/selenium/issues/detail?id=27#c15) originally posted by chrsmith on 2009-06-03T09:57:03.000Z:

Issue 193 has been merged into this issue.

**Comment by chrsmith** [comment]
Comment [#16](https://code.google.com/p/selenium/issues/detail?id=27#c16) originally posted by chrsmith on 2009-06-16T09:52:19.000Z:

We are starting up a testing process with selenium and webdriver.
A solution to this issue would save us a lot of work, because we have a few screens
using alerts/confirmations/modalWindows.

Is there a description of a workaround?

**Comment by chrsmith** [comment]
Comment [#17](https://code.google.com/p/selenium/issues/detail?id=27#c17) originally posted by chrsmith on 2009-06-16T11:58:45.000Z:

Here's what I use to work around the alert problem with the firefox driver.  I don't
have any other modal windows in my app, but I imagine the solution would be similar
for them.   I'm not that happy with it but it does allow me to detect alerts, check
the text and dismiss them.   

Actually it changes what happens when the js alert function is called, so you won't
see an alert box at all.    Also it won't work like a modal alert and block following
messages.  

registerAlertHack() has to be called just once for each page.   However I don't
really know when the page changes in my app, so I call it the whole time in
'sendMessage'.


  driver = new FirefoxDriver()
  {
    protected String sendMessage(Class<? extends RuntimeException> aClass, Command
command)
    {
      String r = super.sendMessage(aClass, command);
      registerAlertHack();
      return r;
    }
  };



private void registerAlertHack()
{
  if (driver != null)
  {
    String js = "if (window.alert.myAlertText == undefined) {window.alert.myAlertText
= null;  window.alert = function(msg){ window.alert.myAlertText = msg; };}";
    ((JavascriptExecutor)driver).executeScript(js);
  }
}

public boolean jsAlertExists()
{
  return  getJsAlertText() != null;
}

public String getJsAlertText()
{
  Object txt = ((JavascriptExecutor)driver).executeScript("return
window.alert.myAlertText;");
  return (String)txt;
}

public void clickJsAlert()
{
  ((JavascriptExecutor)driver).executeScript("window.alert.myAlertText = null;");
}

**Comment by chrsmith** [comment]
Comment [#18](https://code.google.com/p/selenium/issues/detail?id=27#c18) originally posted by chrsmith on 2009-07-26T13:08:11.000Z:

We are starting up a testing process with selenium and webdriver.
A solution to this issue would save us a lot of work, because we have a few screens
using alerts/confirmations/modalWindows.

Please let me know if any updates on this or something which I can implement myself on 
internetexplorerdriver?.

**Comment by chrsmith** [comment]
Comment [#19](https://code.google.com/p/selenium/issues/detail?id=27#c19) originally posted by chrsmith on 2009-08-11T17:44:12.000Z:

I tweaked the workaround that jjimjam submitted to make it a little simpler (for the
JavaScript novice like myself).  In essence you have to know when you will be
generating an alert or confirmation window and you have to know what you want to do
with that window.  So, sometime before the confirmation/alert is generated, just run
this line of code if you want to click OK on a confirm window:
((JavascriptExecutor)m_driver).executeScript("window.confirm = function(msg){return
true;};");

or if you want to click cancel:
((JavascriptExecutor)m_driver).executeScript("window.confirm = function(msg){return
false;};");

or if you want to handle an alert:
((JavascriptExecutor)m_driver).executeScript("window.alert = function(msg){};");

For my situation this seems like a fairly clean situation.  Sure I am overriding the
confirm/alert functions, so they won't handle quite like they would in a real
situation, but I always know before hand what button I want to press on a confirm box
and I really don't care about alert boxes.  Its not a perfect solution, I can't get
the actual message, but its easy and works for me.

**Comment by chrsmith** [comment]
Comment [#20](https://code.google.com/p/selenium/issues/detail?id=27#c20) originally posted by chrsmith on 2010-01-14T16:57:01.000Z:

Hi team, do you have an estimated resolution or implementation date of this issue?

**Comment by chrsmith** [comment]
Comment [#21](https://code.google.com/p/selenium/issues/detail?id=27#c21) originally posted by chrsmith on 2010-03-04T21:43:12.000Z:

I've got a partial fix for the Selenium 1.0.1 version of this bug. With WebDriver and
Selenium merging, I thought I should post it here.

The basic idea is to use the onload event of the iframe containing the AUT to run the
selenium.browserbot._modifyWindow(window) function. 

This fix consists of two changes:
1. Change the RemoteRunner.html file so that this javascript appears in the header:

    function overrideMessageFunctions(iframe) {
        selenium.browserbot._modifyWindow(iframe.contentWindow);
    }
    
    function addHandler()
    {
        var element = window.document.getElementById('selenium_myiframe');
        var listener = function(){ overrideMessageFunctions(element);};
	if (element.addEventListener)
		element.addEventListener("load", listener, false);
	else
		element.attachEvent("onload", listener);
    }

2. In the AUT, change alert() to:
setTimeout(function(){alert();}, 0); 

This solution works in IE8 for sure, no reason why it shouldn't work for others as
well. Attached is the new RemoteRunner.html file.

**Comment by chrsmith** [comment]
Comment [#22](https://code.google.com/p/selenium/issues/detail?id=27#c22) originally posted by chrsmith on 2010-03-07T22:06:49.000Z:

+1 for my vote!

As a workaround, this worked for me, though in my case I did not want a return value
as the modal dialog handled everything for the parent. Anyway thought I'd pass it on.

driver.executeScript("window.showModalDialog = window.openWindow;");

**Comment by chrsmith** [comment]
Comment [#23](https://code.google.com/p/selenium/issues/detail?id=27#c23) originally posted by chrsmith on 2010-03-09T21:24:57.000Z:

Will the resolution for this issue include support for file upload dialogs
(especially in RemoteWebDriver) as well, or should that be opened as a separate issue?

**Comment by chrsmith** [comment]
Comment [#24](https://code.google.com/p/selenium/issues/detail?id=27#c24) originally posted by chrsmith on 2010-04-26T16:10:02.000Z:

Currently using http://yarb.pl/post/505186243/confirming-javascript-alerts-in-capybara 
to effectively stub out the alert/confirm calls.

Goes against the ideals, but it works!

**Comment by chrsmith** [comment]
Comment [#25](https://code.google.com/p/selenium/issues/detail?id=27#c25) originally posted by chrsmith on 2010-05-06T08:12:13.000Z:

It IS a much needed feature...

**Comment by chrsmith** [comment]
Comment [#26](https://code.google.com/p/selenium/issues/detail?id=27#c26) originally posted by chrsmith on 2010-06-17T00:27:46.000Z:

&lt;empty&gt;

**Comment by chrsmith** [comment]
Comment [#27](https://code.google.com/p/selenium/issues/detail?id=27#c27) originally posted by chrsmith on 2010-06-17T09:09:32.000Z:

I have another workaround, which can deal with the alert() before window onload() function, it was bellow:

		FirefoxProfile profile = new FirefoxProfile();
		
		profile.setPreference("capability.policy.policynames", "strict") ;
		profile.setPreference("capability.policy.strict.sites",  "http://61.132.132.70") ;
		profile.setPreference("capability.policy.strict.Window.alert", "noAccess") ;
		profile.setPreference("capability.policy.strict.Window.confirm", "noAccess") ;
		profile.setPreference("capability.policy.strict.Window.prompt", "noAccess") ;
		
		FirefoxDriver driver = new FirefoxDriver(profile);

disable Window.alert() at firefox security preference level.

This solution not good, but seems is the only way I can found now to deal with the alert() before onload(), but only works for firefox ,:(

**Comment by chrsmith** [comment]
Comment [#28](https://code.google.com/p/selenium/issues/detail?id=27#c28) originally posted by chrsmith on 2010-06-22T19:41:57.000Z:

I was looking at the source code and changed the FirefoxTargetLocator class from private to public and the following structure become possible:

Alert alert = driver.switchTo().alert();
alert.accept();

Does anyone know if this work? I couldn't test yet.

**Comment by chrsmith** [comment]
Comment [#29](https://code.google.com/p/selenium/issues/detail?id=27#c29) originally posted by chrsmith on 2010-06-28T19:22:25.000Z:

Issue 590 has been merged into this issue.

**Comment by chrsmith** [comment]
Comment [#30](https://code.google.com/p/selenium/issues/detail?id=27#c30) originally posted by chrsmith on 2010-06-30T06:11:13.000Z:

Really waiting for javascript alerts and prompts/Model Dialog handling

**Comment by chrsmith** [comment]
Comment [#31](https://code.google.com/p/selenium/issues/detail?id=27#c31) originally posted by chrsmith on 2010-07-21T03:10:35.000Z:

This is becoming a blocking issue...i would hate to not able to use selenium because of this.  I can see Firefox code is close, if this works in firefox for start, I would be happy.

**Comment by chrsmith** [comment]
Comment [#32](https://code.google.com/p/selenium/issues/detail?id=27#c32) originally posted by chrsmith on 2010-08-02T07:49:15.000Z:

Hi,

Please make it possible to work with javascript alerts. I am working with watir and would like to use webdriver instead as of its more browsers flexibility.

**Comment by chrsmith** [comment]
Comment [#33](https://code.google.com/p/selenium/issues/detail?id=27#c33) originally posted by chrsmith on 2010-08-05T14:44:23.000Z:

Is there any way at all we could get an ETA on when this feature will become available?

**Comment by chrsmith** [comment]
Comment [#34](https://code.google.com/p/selenium/issues/detail?id=27#c34) originally posted by chrsmith on 2010-08-08T12:42:47.000Z:

I want this feature as much as anyone. It's non-trivial to implement this with the API we want. I would note that this is an OSS project, and if anyone wants to help out and get this one implemented _for any driver_, I'd be happy to accept a patch or to work with them to get this done.

**Comment by chrsmith** [comment]
Comment [#35](https://code.google.com/p/selenium/issues/detail?id=27#c35) originally posted by chrsmith on 2010-08-13T08:04:20.000Z:

Thanks @Santa Li. 
That sounds like a medicine for my headache.

**Comment by chrsmith** [comment]
Comment [#36](https://code.google.com/p/selenium/issues/detail?id=27#c36) originally posted by chrsmith on 2010-08-25T06:17:41.000Z:

This Issue seems to be pending from Oct 2007 & it's a must fix issue.
We need to concentrate on fixing this issue.

**Comment by chrsmith** [comment]
Comment [#37](https://code.google.com/p/selenium/issues/detail?id=27#c37) originally posted by chrsmith on 2010-09-01T17:39:29.000Z:

I would also like this addressed, as we need to verify contents of alert messages occasionally in testing our webapp.  

My workaround was to use Javascript to overwrite the window.alert function to store the message.  After the alert is called, I retrieve the stored message and compare.  So far it works OK with the Firefox and IE drivers.

Below is my workaround.  You will need some extra code to flesh it out, but this shows the main idea.

Java code:

JavascriptExecutor js = (JavascriptExecutor)webDriver;
//store the message
js.executeScript("window.alert = function(msg) { document.lastAlert=msg; }");

//retrieve the alert message
Object o = js.executeScript("return document.lastAlert");

**Comment by chrsmith** [comment]
Comment [#38](https://code.google.com/p/selenium/issues/detail?id=27#c38) originally posted by chrsmith on 2010-09-02T14:13:20.000Z:

#c34

I have been looking at the Firefox code and extended the experimental alerts implementation a little.
I've added tests for handling confirm dialogs, both accept and cancel.
The tests pass, but there is one problem: sometimes they hang.

This does not seem to be the fault of the alert handling though. It hangs after the dialog is closed with the subsequent get request.
(firefoxDriver.js around line 94, the load event is never received). However, adding a few milliseconds wait before the get resolves this.

I'm attaching the patch for your consideration.

**Comment by chrsmith** [comment]
Comment [#39](https://code.google.com/p/selenium/issues/detail?id=27#c39) originally posted by chrsmith on 2010-09-03T14:10:16.000Z:

The Ignores in the patch attached to 38 should be for CHROME, not CHROME_NON_WINDOWS.  Other than that, feels fairly sane, but I'm not hugely familiar with the code being touched

**Comment by chrsmith** [comment]
Comment [#40](https://code.google.com/p/selenium/issues/detail?id=27#c40) originally posted by chrsmith on 2010-09-04T18:25:28.000Z:

Ok, I continued debugging the problem with the previous patch. The response was sent immediately after the click, and it seems that this caused a possible race condition with the subsequent 'get' request.

I added a wait after the click, to check is the click triggered a document load and wait after it's completed to send the response. Now all the tests are passing and I haven't seen hangs anymore. Please review this, since this problem is "not entirely unlike" the normal click code in wrappedElement.js, but that was difficult to refactor because of the anonymous functions in there.

I cleaned up the code a little, also took Daniel's suggestion.
I'm attaching a new patch that supersedes the one from #&nbsp;38.

**Comment by chrsmith** [comment]
Comment [#41](https://code.google.com/p/selenium/issues/detail?id=27#c41) originally posted by chrsmith on 2010-09-09T13:20:38.000Z:

Hi Guys,
I'm using Watir-Webdriver and RUBY.
I am currently automating a CMS and I've come across a dialog box which I just have to click on OK and then I can continue with my test automation. BUT I am having a lot of difficulty trying to click on the OK button in the dialog pop up.

I've used UI-SPY to get the attributes of the OK button (screen shot has been attached).

Is there anyone who has done this using Watir-Webdriver?
Any help will be brilliant. This is turning into a Blocker for me. If you need any more details please let me know.

Thanks in advance

Kind Regards,
Usman

**Comment by chrsmith** [comment]
Comment [#42](https://code.google.com/p/selenium/issues/detail?id=27#c42) originally posted by chrsmith on 2010-09-09T13:54:26.000Z:

@usmanhhussain

If the window is created in JS using alert(), prompt() or confirm(), check out this workaround: 

http://stackoverflow.com/questions/3652582/alert-box-in-waitr-webdriver

If not, please provide more details.

**Comment by chrsmith** [comment]
Comment [#43](https://code.google.com/p/selenium/issues/detail?id=27#c43) originally posted by chrsmith on 2010-09-09T14:40:02.000Z:

how can i see or tell if the window is created in Javascript?

kind regards,
Usman Hussain

**Comment by chrsmith** [comment]
Comment [#44](https://code.google.com/p/selenium/issues/detail?id=27#c44) originally posted by chrsmith on 2010-09-09T16:25:30.000Z:

Ive tried the link you provided but still nothing. keep getting an error.

Ive read up about autoit and tried to install the dll manually but it does not install. is there anyway i can get autoit to be installed so that it works with Watir-WebDriver?

The reason for me doing this is so that i can then just do a keystroke command for Enter...

Kind regards,
Usman Hussain

**Comment by chrsmith** [comment]
Comment [#45](https://code.google.com/p/selenium/issues/detail?id=27#c45) originally posted by chrsmith on 2010-09-09T16:59:08.000Z:

@usmanhhussain: I'll follow up in your thread on wtr-general.

**Comment by chrsmith** [comment]
Comment [#46](https://code.google.com/p/selenium/issues/detail?id=27#c46) originally posted by chrsmith on 2010-10-08T22:30:44.000Z:

Hi Simon

I have implemented a fix for this Alerts issue into the Internet Explorer Driver with the desired API 
Alert alert = driver.switchTo().alert().accept
Alert alert = driver.switchTo().alert().dismiss
I have uploaded the Internet explorer driver jar and some test jsp files.
Apologies please use jdk6.

If anyone else would like to grab the driver and test it out and give me some feedback that would be really great.

Many thanks

Chi

**Comment by chrsmith** [comment]
Comment [#47](https://code.google.com/p/selenium/issues/detail?id=27#c47) originally posted by chrsmith on 2010-10-09T05:57:36.000Z:

How to handle this in Selenium 2.0 in C#

Anybody could help me a little bit?

**Comment by chrsmith** [comment]
Comment [#48](https://code.google.com/p/selenium/issues/detail?id=27#c48) originally posted by chrsmith on 2010-10-10T20:46:19.000Z:

Hi Binken Zou,

I have added the C# bindings for the IE (Internet Explorer) driver so you should be able to test my fix with 
C# rather than Java (the IE driver is actually implemented in C++ of course).
Please give it a go and let me know (i'd welcome any feedback as to issues with my potential fix)

Cheers


Chi

**Comment by chrsmith** [comment]
Comment [#49](https://code.google.com/p/selenium/issues/detail?id=27#c49) originally posted by chrsmith on 2010-10-11T15:47:16.000Z:

Hey Chi,

For some reason InternetExplorerDriver.dll doesnt load into my references, and WebDriver.Common and .IE wont run without it.

**Comment by chrsmith** [comment]
Comment [#50](https://code.google.com/p/selenium/issues/detail?id=27#c50) originally posted by chrsmith on 2010-10-11T17:12:35.000Z:

Hi Jim

First of all thks for downloading.  The InternetExplorerDriver.dll is not a COM DLL
and is a straight DLL (the actual IE driver itself).  This file just needs to be in your directory (some where in your build/bin directory etc).
Please give that a go if you still have probs i'll send you a zip of my visual studio 2008 test client project (which tests the IE driver fix).

many regards


Chi

**Comment by chrsmith** [comment]
Comment [#51](https://code.google.com/p/selenium/issues/detail?id=27#c51) originally posted by chrsmith on 2010-10-11T23:13:20.000Z:

Hi Jim

First of all thks for downloading.  The InternetExplorerDriver.dll is not a COM DLL
and is a straight DLL (the actual IE driver itself).  This file just needs to be in your directory (some where in your build/bin directory)
Apologies i'm being a bit vague with this i did the .NET bindings in a rush
and just copied the DLL into the bin/debug directory of the test project and it just seems to have picked up.
Apologies for being vague once again, the Java version just has everything bundled in a single JAR file and you just reference that and it works.
Sorry i have not had time to look into precisely how that DLL gets picked up for .NET as i just wanted to get something up on the site as i have had the fix for quite a while now and thought it was really about time i fitted it into the actual IE driver code itself ie. the C++ InternetExplorerDriver.dll. 
I have tested the fix on both C++, Java, CSharp test client projects and it accepts and dismisses Alerts whether they may appear in the immediately load of a page, on click of a button etc etc.  When applying the fix into the driver i have been careful to make no assumptions as to the timing of the appearance of the actual Alert.
This is to maintain that great blocking API which i believe is crucial for a much more robust reliable solution.
As a result you should find that eg code line like 

Alert myAlert = ieDriver.switchTo().Alert();

will block until the actual driver gets a hook onto the Alert box, so that by the time you are onto your next line of code
eg.

myAlert.accept(); or myAlert.dismiss()

the Alert is there waiting for you to accept or dismiss.

I have explicitly tested for this by eg. clicking on a button that then navigates to another page and that page 
takes a while to load before and Alert box gets raised.
The driver will wait for the Alert to show before moving on from the 

ieDriver.switchTo().Alert().accept()

code line.

If you run the same test with the existing switch to window API 

ie. ieDriver.switchTo().Window();

you will find that it behaves differently in that it does not actually block and consequently if your next line of code tries to find an element on the switched to window it will fail.  Not sure if this has been raised by anyone but will find out.

Sorry for being so verbose i thought i would put down some of my thinking (so that others may comment in terms of ideas, issues etc etc in the spirit of open source)
I think this is fantastic open source project and would really like to see this issue fixed.

Strange/a little ironic i have not had am Java people trying the fix yet ? (i'm more Java than CSharp)


Please give that a go if you still have probs i'll send you a zip of my visual studio 2008 test client project (which tests the IE driver fix).

many regards


Chi

**Comment by chrsmith** [comment]
Comment [#52](https://code.google.com/p/selenium/issues/detail?id=27#c52) originally posted by chrsmith on 2010-10-21T16:02:05.000Z:

Hi basically I am trying to switch to a popup window then press a
button on that popup window but from somereaon I am getting erros.

Here is an example of the webpage

http://i42.photobucket.com/albums/e326/limpep/example.png

and here is my code

 for (String handle : driver.getWindowHandles()) {
                                driver.switchTo().window(handle);

                              }


                      WebElement clicksa  =
driver.findElement(By.id("ButtonCancle"));
                          clicksa.sendKeys(Keys.ENTER);
                          driver.switchTo().defaultContent();

Errers that i am getting

   Exception in thread "AWT-EventQueue-0"
org.openqa.selenium.NoSuchElementException: Unable to find element by
id using "Accept Address" (7)
System info: os.name: 'Windows XP', os.arch: 'x86', os.version: '5.1',
java.version: '1.6.0_21'
Driver info: driver.version: ie
   at
org.openqa.selenium.ie.ErrorHandler.verifyErrorCode(ErrorHandler.java:
38)
   at org.openqa.selenium.ie.Finder.handleErrorCode(Finder.java:252)
   at org.openqa.selenium.ie.Finder.findElementById(Finder.java:126)
   at org.openqa.selenium.By$1.findElement(By.java:66)
   at org.openqa.selenium.ie.Finder.findElement(Finder.java:240)
   at
org.openqa.selenium.ie.InternetExplorerDriver.findElement(InternetExplorerDriver.java:
297)
   at com.capscanWebServers.CRMTesting
$2.actionPerformed(CRMTesting.java:112)
   at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
   at javax.swing.AbstractButton$Handler.actionPerformed(Unknown
Source)
   at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown
Source)
   at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
   at
javax.swing.plaf.basic.BasicButtonListener.mouseReleased(Unknown
Source)
   at java.awt.Component.processMouseEvent(Unknown Source)
   at javax.swing.JComponent.processMouseEvent(Unknown Source)
   at java.awt.Component.processEvent(Unknown Source)
   at java.awt.Container.processEvent(Unknown Source)
   at java.awt.Component.dispatchEventImpl(Unknown Source)
   at java.awt.Container.dispatchEventImpl(Unknown Source)
   at java.awt.Component.dispatchEvent(Unknown Source)
   at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown
Source)
   at java.awt.LightweightDispatcher.processMouseEvent(Unknown
Source)
   at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
   at java.awt.Container.dispatchEventImpl(Unknown Source)
   at java.awt.Window.dispatchEventImpl(Unknown Source)
   at java.awt.Component.dispatchEvent(Unknown Source)
   at java.awt.EventQueue.dispatchEvent(Unknown Source)
   at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
   at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
   at java.awt.EventDispatchThread.run(Unknown Source)

**Comment by chrsmith** [comment]
Comment [#53](https://code.google.com/p/selenium/issues/detail?id=27#c53) originally posted by chrsmith on 2010-10-21T16:03:07.000Z:

Hi basically I am trying to switch to a popup window then press a
button on that popup window but from somereaon I am getting erros.

Here is an example of the webpage

http://i42.photobucket.com/albums/e326/limpep/example.png

and here is my code

 for (String handle : driver.getWindowHandles()) {
                                driver.switchTo().window(handle);

                              }


                      WebElement clicksa  =
driver.findElement(By.id("Accept Address"));
                          clicksa.sendKeys(Keys.ENTER);
                          driver.switchTo().defaultContent();

Errers that i am getting

   Exception in thread "AWT-EventQueue-0"
org.openqa.selenium.NoSuchElementException: Unable to find element by
id using "Accept Address" (7)
System info: os.name: 'Windows XP', os.arch: 'x86', os.version: '5.1',
java.version: '1.6.0_21'
Driver info: driver.version: ie
   at
org.openqa.selenium.ie.ErrorHandler.verifyErrorCode(ErrorHandler.java:
38)
   at org.openqa.selenium.ie.Finder.handleErrorCode(Finder.java:252)
   at org.openqa.selenium.ie.Finder.findElementById(Finder.java:126)
   at org.openqa.selenium.By$1.findElement(By.java:66)
   at org.openqa.selenium.ie.Finder.findElement(Finder.java:240)
   at
org.openqa.selenium.ie.InternetExplorerDriver.findElement(InternetExplorerDriver.java:
297)
   at com.capscanWebServers.CRMTesting
$2.actionPerformed(CRMTesting.java:112)
   at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
   at javax.swing.AbstractButton$Handler.actionPerformed(Unknown
Source)
   at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown
Source)
   at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
   at
javax.swing.plaf.basic.BasicButtonListener.mouseReleased(Unknown
Source)
   at java.awt.Component.processMouseEvent(Unknown Source)
   at javax.swing.JComponent.processMouseEvent(Unknown Source)
   at java.awt.Component.processEvent(Unknown Source)
   at java.awt.Container.processEvent(Unknown Source)
   at java.awt.Component.dispatchEventImpl(Unknown Source)
   at java.awt.Container.dispatchEventImpl(Unknown Source)
   at java.awt.Component.dispatchEvent(Unknown Source)
   at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown
Source)
   at java.awt.LightweightDispatcher.processMouseEvent(Unknown
Source)
   at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
   at java.awt.Container.dispatchEventImpl(Unknown Source)
   at java.awt.Window.dispatchEventImpl(Unknown Source)
   at java.awt.Component.dispatchEvent(Unknown Source)
   at java.awt.EventQueue.dispatchEvent(Unknown Source)
   at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown
Source)
   at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
   at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
   at java.awt.EventDispatchThread.run(Unknown Source)

**Comment by chrsmith** [comment]
Comment [#54](https://code.google.com/p/selenium/issues/detail?id=27#c54) originally posted by chrsmith on 2010-10-22T14:27:05.000Z:

Chi, this sounds great. Are there any plans to make a similar fix for showModalDialog? Seems similar, a blocking API that pops up a window that test code needs to manipulate. Thanks
Tom

**Comment by chrsmith** [comment]
Comment [#55](https://code.google.com/p/selenium/issues/detail?id=27#c55) originally posted by chrsmith on 2010-10-25T12:31:00.000Z:

Hi Tom

I'm not sure about immediately looking into something like showModalDialog just yet
. I thought this was an interesting blocking issue to have a stab at first.
However you're right about something like that possible being along the sames lines.
Is there a separate logged issue for this showModalDialog stuff ?
By the way if you want to download and try the fix i have not uploaded the jna.jar which is the 3rd party jar which the IE driver uses to communicate between the Java and C++.  You can just google for that.

Many Regards


Chi

**Comment by chrsmith** [comment]
Comment [#56](https://code.google.com/p/selenium/issues/detail?id=27#c56) originally posted by chrsmith on 2010-10-25T13:33:15.000Z:

Yes, issue 284 tracks showModalDialog. There are workarounds that are sometimes effective, but a real fix would be useful.

Thanks!
Tom

**Comment by chrsmith** [comment]
Comment [#57](https://code.google.com/p/selenium/issues/detail?id=27#c57) originally posted by chrsmith on 2010-12-01T22:17:42.000Z:

Beta1 will have an implementation of the Alerts and Prompts API for the FirefoxDriver.

**Comment by chrsmith** [comment]
Comment [#58](https://code.google.com/p/selenium/issues/detail?id=27#c58) originally posted by chrsmith on 2010-12-08T07:31:27.000Z:

Hope it will not only available in Java

**Comment by chrsmith** [comment]
Comment [#59](https://code.google.com/p/selenium/issues/detail?id=27#c59) originally posted by chrsmith on 2010-12-09T12:38:42.000Z:

@binken.zou: when its done in Java it will be ported around

**Comment by chrsmith** [comment]
Comment [#60](https://code.google.com/p/selenium/issues/detail?id=27#c60) originally posted by chrsmith on 2010-12-09T13:13:30.000Z:

I added support in the Ruby bindings (for Firefox/Remote) in r10585. 

I had to downgrade to Firefox 3 to get the tests passing though - the current Firefox alert/prompt implementation doesn't seem to work with the Firefox 4 beta.

**Comment by chrsmith** [comment]
Comment [#61](https://code.google.com/p/selenium/issues/detail?id=27#c61) originally posted by chrsmith on 2010-12-13T17:25:36.000Z:

Here's the failure on Fx b6 (can't test beta 7 on my Leopard machine atm):

https://gist.github.com/739196

**Comment by chrsmith** [comment]
Comment [#62](https://code.google.com/p/selenium/issues/detail?id=27#c62) originally posted by chrsmith on 2010-12-23T10:05:23.000Z:

I'm going to test MS Share point with NTLM Authentication scheme. What can I do in this situation?

**Comment by chrsmith** [comment]
Comment [#63](https://code.google.com/p/selenium/issues/detail?id=27#c63) originally posted by chrsmith on 2010-12-31T06:39:06.000Z:

Hi Chi,
Advance Happy New Year to you and all...
I am trying out Selenium 2 with C# and have a query here;
a. How can I handle File Download Window dialog and press Save button as shown in snapshot (A)
b. How can I enter the file path and then press Save button as shown in snapshot (B)

I tried with >>
IAlert ActionElement1 = driver.SwitchTo().Alert();
ActionElement1.SendKeys(Keys.Alt+"S");

The above code does not work and really frustrating.
I have used AutoIT to handle the similar scenario with Selenium 1.0 but looking to eliminate the use of AutoIT.

Let me know if this can be done.

Thank you,
Yeshwant

**Comment by chrsmith** [comment]
Comment [#64](https://code.google.com/p/selenium/issues/detail?id=27#c64) originally posted by chrsmith on 2010-12-31T16:46:57.000Z:

The alert API is designed for use ONLY with JavaScript alerts. That is, dialog boxes appearing through the use of the JavaScript alert(), confirm(), and prompt() functions. Other alerts (such as file selection dialogs) will not work at the moment.

**Comment by chrsmith** [comment]
Comment [#65](https://code.google.com/p/selenium/issues/detail?id=27#c65) originally posted by chrsmith on 2011-01-04T21:12:22.000Z:

There appears be issues with the 2.0 Beta 1 alert API if the alert is generated by a Web Element that is embedded within a frame.  

I have successfully verified interaction with a sample javascript alert directly against the html source page.  When the page is embedded in a frame, the firefox driver blocking implementation hangs the test until the alert is accepted or closed manually.

Are there any known workarounds for this behavior?

**Comment by chrsmith** [comment]
Comment [#66](https://code.google.com/p/selenium/issues/detail?id=27#c66) originally posted by chrsmith on 2011-01-10T18:17:02.000Z:

I have the same problem. I click on an item within an iframe that should display a confirmation, but click method hangs, and waits until i click ok, or cancel.

**Comment by chrsmith** [comment]
Comment [#67](https://code.google.com/p/selenium/issues/detail?id=27#c67) originally posted by chrsmith on 2011-01-12T22:56:30.000Z:

Issue 1191 has been merged into this issue.

**Comment by chrsmith** [comment]
Comment [#68](https://code.google.com/p/selenium/issues/detail?id=27#c68) originally posted by chrsmith on 2011-02-01T17:47:20.000Z:

+1 for the issue regarding alerts generated by elements within an iframe.  I'm using 2.0b1 and seeing the exact same behavior.  It's similar to the behavior I see when an iframe element generates a new modal window.  The click() method hangs until the modal is closed.

**Comment by chrsmith** [comment]
Comment [#69](https://code.google.com/p/selenium/issues/detail?id=27#c69) originally posted by chrsmith on 2011-02-08T11:32:58.000Z:

Hi,
 how to handle web page authentication prompts in selenium Web Drivers.
(For eg.https://members.aesc.org/iweb/). 

thanks..
Rajesh

**Comment by chrsmith** [comment]
Comment [#70](https://code.google.com/p/selenium/issues/detail?id=27#c70) originally posted by chrsmith on 2011-02-11T09:16:18.000Z:

Hi,
How can I get messages in confirm window?
thanks.

**Comment by chrsmith** [comment]
Comment [#71](https://code.google.com/p/selenium/issues/detail?id=27#c71) originally posted by chrsmith on 2011-02-18T17:40:36.000Z:

I created a separate issue for the problem with alerts/modal dialogs that appear as a result of interacting with an element in an iframe.

http://code.google.com/p/selenium/issues/detail?id=1305

**Comment by chrsmith** [comment]
Comment [#72](https://code.google.com/p/selenium/issues/detail?id=27#c72) originally posted by chrsmith on 2011-04-05T04:33:07.000Z:

The Alert does not seem to work with Firefox4.0 :-(

**Comment by chrsmith** [comment]
Comment [#73](https://code.google.com/p/selenium/issues/detail?id=27#c73) originally posted by chrsmith on 2011-04-07T16:05:35.000Z:

Same problem here. Firefox4, selenium2b3
Trying to accept an alert that is launched from a frame. But the script hangs on the click command that launches the frame, until I manually accept it.

**Comment by chrsmith** [comment]
Comment [#74](https://code.google.com/p/selenium/issues/detail?id=27#c74) originally posted by chrsmith on 2011-04-07T20:27:45.000Z:

I'm also having trouble with Alert with Firefox 4.0, using the 2.0b3 .NET bindings.

WebDriver.SwitchTo().Alert() finds an alert just fine, but calling .Accept() on the the Alert object causes a WebDriverException: "Unexpected error. Unable to dismiss alert. Is an alert present?".

**Comment by chrsmith** [comment]
Comment [#75](https://code.google.com/p/selenium/issues/detail?id=27#c75) originally posted by chrsmith on 2011-04-09T10:47:03.000Z:

I'm also having issues with Alerts. I'm using the Ruby bindings, so I thinks it an issue of selenium core.

**Comment by chrsmith** [comment]
Comment [#76](https://code.google.com/p/selenium/issues/detail?id=27#c76) originally posted by chrsmith on 2011-04-19T02:59:43.000Z:

Same here. Need to test a webpage. A popup appears asking for OK or Cancel and cant do much. Testing with Selenium2.0b3 with the C# driver. Tested with the python version and same issue is observed which makes it to be something related to firefox and selenium implementation. Would be great if this issue was solved as soon as possible as a lot of people are waiting for the fix.

**Comment by chrsmith** [comment]
Comment [#77](https://code.google.com/p/selenium/issues/detail?id=27#c77) originally posted by chrsmith on 2011-04-19T17:35:48.000Z:

+1 the same issue as  poddarsu...@gmail.com,

**Comment by chrsmith** [comment]
Comment [#78](https://code.google.com/p/selenium/issues/detail?id=27#c78) originally posted by chrsmith on 2011-04-20T07:57:20.000Z:

I don't have a solution, just working workaround (tested only on Firefox 3.6 and MSIE 8). It uses JavaScript, so I suppose it would work in any supported langauge, I'm using only Java. See http://rostislav-matl.blogspot.com/2011/04/moving-to-selenium-2-on-webdriver-part.html if interested.

**Comment by chrsmith** [comment]
Comment [#79](https://code.google.com/p/selenium/issues/detail?id=27#c79) originally posted by chrsmith on 2011-04-21T14:18:39.000Z:

Hi give the IE driver implementation of the alerts API a go with the same tests, does that work ? or are there issues ?

Many regards

**Comment by chrsmith** [comment]
Comment [#80](https://code.google.com/p/selenium/issues/detail?id=27#c80) originally posted by chrsmith on 2011-04-22T16:20:30.000Z:

Alerts and prompts have been supported in the native code portion of the IE driver since r10866 (06 January 2011).

**Comment by chrsmith** [comment]
Comment [#81](https://code.google.com/p/selenium/issues/detail?id=27#c81) originally posted by chrsmith on 2011-04-24T14:43:07.000Z:

Hi Jim

Sorry it has indeed been several months
hope you are well

Chi

**Comment by chrsmith** [comment]
Comment [#82](https://code.google.com/p/selenium/issues/detail?id=27#c82) originally posted by chrsmith on 2011-05-10T13:19:42.000Z:

Im having a problem accepting an alert in FF4 with C# (Selenium WebDriver) 2.0b3

**Comment by chrsmith** [comment]
Comment [#83](https://code.google.com/p/selenium/issues/detail?id=27#c83) originally posted by chrsmith on 2011-05-12T00:02:29.000Z:

I'm attaching a patch which fixes the described problem with modal window.confirm appearing in Firefox 4.0.1, Selenium 2.0b3, Linux and delaying test execution until the popup is discarded manually. This is a regression that I observed after upgrading from Firefox 3.6.x and Selenium RC 1.0.3.

If you wish to try it out, extract the contents of selenium-server-standalone-2.0b3.jar, apply the patch to file core/scripts/selenium-browserbot.js, then update the jar like so:

jar umf META-INF/MANIFEST.MF selenium-server-standalone-2.0b3.jar core/scripts/selenium-browserbot.js

I don't know why this patch works. There is something strange going on with the overridden confirm function (for example). In the original version of browserbot.js if you try the immediate sequence of calls windowToModify.confirm('first test'); windowToModify.confirm('second test'); right after assigning the new function to windowToModify.confirm, then you will notice that the first call goes to the new function, but the second call goes to the native (original) function. This occurs regardless of the content of the new function - the reference to the original function is somehow magically restored.

**Comment by chrsmith** [comment]
Comment [#84](https://code.google.com/p/selenium/issues/detail?id=27#c84) originally posted by chrsmith on 2011-05-12T09:46:45.000Z:

jpljpl, great work!

**Comment by chrsmith** [comment]
Comment [#85](https://code.google.com/p/selenium/issues/detail?id=27#c85) originally posted by chrsmith on 2011-05-22T07:57:59.000Z:

It seems this issue still affects FF4 + latest trunk (r12228), and jpl++'s patch is still required to work -- would it be sensible to merge it as part of 2.0b4?

Cheers,
Audrey

**Comment by chrsmith** [comment]
Comment [#86](https://code.google.com/p/selenium/issues/detail?id=27#c86) originally posted by chrsmith on 2011-05-22T11:24:02.000Z:

If the person merging it, unlike jpl++, actually knows WHY it works! :)

**Comment by chrsmith** [comment]
Comment [#87](https://code.google.com/p/selenium/issues/detail?id=27#c87) originally posted by chrsmith on 2011-06-07T16:31:07.000Z:

Using the 2.0rc2 release.
The following code snippet inside a Selenium test throws a NullPointerException:

		selenium.click("//input[@id='']");
		selenium.waitForPageToLoad("30000");
		Alert alert = driver.switchTo().alert();
		selenium.waitForPageToLoad("30000");
assertTrue(alert.getText().matches("^Confirmez-vous votre choix [\\s\\S]$")); // alert == null
		alert.accept();

When not running the test, clicking the button (1st instruction) triggers a confirmation box with expected title.
When running the test, the confirmation popup is not visible as expected, but alert is null.
If I remove the offending assertion, then I get a WebDriverException: Unable to dismiss alert. Is alert present?

What am I doing wrong? I assumed the initialisation of alert would block waiting for an alert to come. I couldn't find a waitForAlert() or similar call.
How should I code to get the test to pass, checking that a confirmation pops up with expected text?

Many thanks,
CChris

**Comment by chrsmith** [comment]
Comment [#88](https://code.google.com/p/selenium/issues/detail?id=27#c88) originally posted by chrsmith on 2011-07-06T14:37:46.000Z:

I'm also having some issues with working with alerts. I'm using 2.0rc3 and I want to check if the alert is present or not before trying to interact with it. However I cannot seem to find any API for this. 
Thanks. 
Eugen.

**Comment by chrsmith** [comment]
Comment [#89](https://code.google.com/p/selenium/issues/detail?id=27#c89) originally posted by chrsmith on 2011-07-06T18:04:51.000Z:

Well, you could do something like switching to the alert, and catching the exception that will be thrown if it doesn't exist.
But I think this is a question for the group: http://groups.google.com/group/selenium-users
Please send it there.

**Comment by chrsmith** [comment]
Comment [#90](https://code.google.com/p/selenium/issues/detail?id=27#c90) originally posted by chrsmith on 2011-07-07T09:03:29.000Z:

Yes, I considered this solution, but it didn't seem to me as the way to go if another option is available. I will post on the list. Thanks.

**Comment by chrsmith** [comment]
Comment [#91](https://code.google.com/p/selenium/issues/detail?id=27#c91) originally posted by chrsmith on 2011-07-14T14:48:37.000Z:

I really think this should be changed from enhancement to defect, now that FF4 has been out in the wild for so long.

**Comment by chrsmith** [comment]
Comment [#92](https://code.google.com/p/selenium/issues/detail?id=27#c92) originally posted by chrsmith on 2011-07-15T04:48:17.000Z:

Alert handling does work for FF4 in both WebDriver and Selenium RC (and for FF5 it does work too). Chrome is the browser that is not supported yet. If you have any particular issue with alerts in FF, please, describe it in more details.

**Comment by chrsmith** [comment]
Comment [#93](https://code.google.com/p/selenium/issues/detail?id=27#c93) originally posted by chrsmith on 2011-07-26T11:56:25.000Z:

The Alert handling does not work in FF3.6.

**Comment by chrsmith** [comment]
Comment [#94](https://code.google.com/p/selenium/issues/detail?id=27#c94) originally posted by chrsmith on 2011-07-26T22:46:57.000Z:

What version ? What scenario ? Last time I checked it worked (in RC02, RC03).

**Comment by chrsmith** [comment]
Comment [#95](https://code.google.com/p/selenium/issues/detail?id=27#c95) originally posted by chrsmith on 2011-07-27T06:23:29.000Z:

Version: 2.0.0
Setup: Client/Server

Specific issue: Confirmation 

Server Setup:  
java -jar selenium-server-standalone-2.0.0.jar -role hub -host localhost -port 4444
Java –jar selenium-server-standalone-2.0.0.jar –role webdriver -hub http://localhost:4444/grid/register -host localhost -port 5555 -browser browserName=firefox,version=3.6,maxInstances=3,platform=WINDOWS

Pom.xml:
<dependency>
  <groupId>org.seleniumhq.selenium</groupId>
  <artifactId>selenium-java</artifactId>
  <version>2.2.0</version>
  <scope>test</scope>
</dependency>

Java Client Setup:
DesiredCapabilities capability = DesiredCapabilities.firefox();
capability.setPlatform(Platform.WINDOWS);
capability.setVersion("3.6");
WebDriver webDriver = new RemoteWebDriver(new URL("http://localhost:4444/wd/hub"), capability);
Selenium selenium = new WebDriverBackedSelenium(webDriver, "http://localhost:8080/portal");

// Code being used
public String getConfirmation() {
   String message;
   try {
      Alert alert = webDriver.switchTo().alert();
      message = alert.getText();
      alert.accept();
   } catch (final WebDriverException e) {
      message = null;
   }
   return message;
}

Description: Running the test, the javascript confirmation popup never appears and the message retuned is null. But manually repeating the test it does appear.

Running the same automated test against FF4 or FF5, the popup appears and the message is returned correctly.

**Comment by chrsmith** [comment]
Comment [#96](https://code.google.com/p/selenium/issues/detail?id=27#c96) originally posted by chrsmith on 2011-07-27T06:24:18.000Z:

Version: 2.0.0
Setup: Client/Server

Specific issue: Confirmation 

Server Setup:  
java -jar selenium-server-standalone-2.0.0.jar -role hub -host localhost -port 4444
Java –jar selenium-server-standalone-2.0.0.jar –role webdriver -hub http://localhost:4444/grid/register -host localhost -port 5555 -browser browserName=firefox,version=3.6,maxInstances=3,platform=WINDOWS

Pom.xml:
<dependency>
  <groupId>org.seleniumhq.selenium</groupId>
  <artifactId>selenium-java</artifactId>
  <version>2.0.0</version>
  <scope>test</scope>
</dependency>

Java Client Setup:
DesiredCapabilities capability = DesiredCapabilities.firefox();
capability.setPlatform(Platform.WINDOWS);
capability.setVersion("3.6");
WebDriver webDriver = new RemoteWebDriver(new URL("http://localhost:4444/wd/hub" capability);
Selenium selenium = new WebDriverBackedSelenium(webDriver, "http://localhost:8080/portal");

// Code being used
public String getConfirmation() {
   String message;
   try {
      Alert alert = webDriver.switchTo().alert();
      message = alert.getText();
      alert.accept();
   } catch (final WebDriverException e) {
      message = null;
   }
   return message;
}

Description: Running the test, the javascript confirmation popup never appears and the message retuned is null. But manually repeating the test it does appear.

Running the same automated test against FF4 or FF5, the popup appears and the message is returned correctly.

**Comment by chrsmith** [comment]
Comment [#97](https://code.google.com/p/selenium/issues/detail?id=27#c97) originally posted by chrsmith on 2011-07-28T15:00:34.000Z:

I have this working without problem, Selenium 2.2, Firefox 3.6.15:

  driver.get("http://www.echoecho.com/samplecookie1.htm");
  Alert prompt = driver.switchTo().alert();
  WebDriver.Options driverOptions = driver.manage();
  driverOptions.addCookie( new Cookie("testCookie","BBB") );
  //waitSeconds(1);
  out.println(prompt.getText());
  prompt.sendKeys("AAA");
  //waitSeconds(1);
  prompt.accept();
  //waitSeconds(2);
  Set<Cookie> cookies = driverOptions.getCookies();
  for(Cookie cookie : cookies) 
  {out.println("COOKIE: " + cookie.getName() + ", " + cookie.getValue() ); }

**Comment by chrsmith** [comment]
Comment [#98](https://code.google.com/p/selenium/issues/detail?id=27#c98) originally posted by chrsmith on 2011-08-03T08:16:59.000Z:

Alerts work for java FirefoxDriver 2.3.0, but not for IE 8 nor Chrome on XP. Sorry, this should be made critical issue as alert messages are basic functionality of any web page. I had to use workaround with javascript as in comment #&nbsp;78 and it works but it's not pretty http://code.google.com/p/selenium/issues/detail?id=27#c78
Also, driver.switchTo().alert() should return null, not throw Exception. The problem is that many times it does not even get to this statment if you have driver.submit() and javascript validation code that results in alert window and execution hangs on submit until one clicks on ok button. The work around in #&nbsp;78 works for that scenario, so before submit alert needs to be disabled and after enabled.

**Comment by chrsmith** [comment]
Comment [#99](https://code.google.com/p/selenium/issues/detail?id=27#c99) originally posted by chrsmith on 2011-08-26T08:10:42.000Z:

Am I right to think that there is not any way to handle JS alerts invoked in page onload? Such alerts hang up an WebDriver and tests are stopped.

**Comment by chrsmith** [comment]
Comment [#100](https://code.google.com/p/selenium/issues/detail?id=27#c100) originally posted by chrsmith on 2011-08-26T17:00:15.000Z:

@99: At the moment, yes

**Comment by chrsmith** [comment]
Comment [#101](https://code.google.com/p/selenium/issues/detail?id=27#c101) originally posted by chrsmith on 2011-09-01T12:36:57.000Z:

Any way to handle windows security popups? alert.SendKeys() can send strings to user name textbox, but cannot send special keys like TAB or ENTER. And alert.Accept() does not work either, with an exception saying something like "could not find OK button"...

**Comment by chrsmith** [comment]
Comment [#102](https://code.google.com/p/selenium/issues/detail?id=27#c102) originally posted by chrsmith on 2011-09-12T23:49:30.000Z:

when calling Selenium::WebDriver.get to load a page with an alert, while debuging i noticed the application hangs until a timeout is reached. Then it just throws an error. If I can suggest anything, an event callback should be implemented: driver.onAlert = myAlertHandler, etc. As opposed to injecting some JavaScript wich apparently only works in some cases. On the other hand calling Selenium::WebDriver.navigate.to instead went over to the next call in the stack wich game me hope but then when calling driver.switch_to.alert.dismiss , it hung again untill the thrown timeout error. im using a java remote server and selenium webdriver 2.5.0. with firefox.

**Comment by chrsmith** [comment]
Comment [#103](https://code.google.com/p/selenium/issues/detail?id=27#c103) originally posted by chrsmith on 2011-11-22T17:50:53.000Z:

All drivers, to some degree, now support JS alerts and prompts. There are other issues to do with security dialogs and similar, but those are different from this. Closing this issue.

**Comment by chrsmith** [comment]
Comment [#104](https://code.google.com/p/selenium/issues/detail?id=27#c104) originally posted by chrsmith on 2011-11-28T14:57:51.000Z:

Is this not fixed for chrome driver? From where I can download the new driver?

**Comment by chrsmith** [comment]
Comment [#105](https://code.google.com/p/selenium/issues/detail?id=27#c105) originally posted by chrsmith on 2012-03-27T05:44:07.000Z:

I was using this command when using require 'test/unit'

$b.confirm(true) do
      $b.button(:value,"reload").when_present.fire_event'onclick' # Reset the Database
end

It works perfectly in that case to handle the java script pop up which come up when I click on the button named "reload"..

now I have switched to using require "rspec" instead and some how if now I use the same command I get following error

 undefined method `confirm' for nil:NilClass

could some one help me in fixing it

**Comment by chrsmith** [comment]
Comment [#106](https://code.google.com/p/selenium/issues/detail?id=27#c106) originally posted by chrsmith on 2012-08-28T21:51:28.000Z:

This doesn't seem to be fixed. Not even close!!! I would recommend re-opening this issue.....

PS, it seems that just clicking something which causes a modal dialog to appear also causes issues.

**Comment by chrsmith** [comment]
Comment [#107](https://code.google.com/p/selenium/issues/detail?id=27#c107) originally posted by chrsmith on 2013-01-22T20:23:00.000Z:

The title of this issue is rather broad.  It was mentioned in this thread about handling alerts that occur in the onload of a page.  Is it possible to handle that scenario with WebDriver?  If not can a new issue be opened for that specific issue?

**Comment by chrsmith** [comment]
Comment [#108](https://code.google.com/p/selenium/issues/detail?id=27#c108) originally posted by chrsmith on 2013-01-22T21:05:37.000Z:

@JimRAnderson WebDriver already supports this. There are tests in the WebDriver test suite that test this[1], and the project's CI server[2] executes them for every commit.

[1] http://code.google.com/p/selenium/source/browse/java/client/test/org/openqa/selenium/AlertsTest.java#&nbsp;333
[2] http://ci.seleniumhq.org:8080/

**Comment by chrsmith** [comment]
Comment [#109](https://code.google.com/p/selenium/issues/detail?id=27#c109) originally posted by chrsmith on 2013-04-24T10:41:43.000Z:

I am getting modalwindow error,when i click a button.Alert is getting suppressed.
It stays for few seconds and modalWindow Exception comesUp.

**Comment by chrsmith** [comment]
Comment [#110](https://code.google.com/p/selenium/issues/detail?id=27#c110) originally posted by chrsmith on 2013-04-24T10:46:07.000Z:

&lt;empty&gt;

**Comment by chrsmith** [comment]
Comment [#111](https://code.google.com/p/selenium/issues/detail?id=27#c111) originally posted by chrsmith on 2013-04-24T10:46:58.000Z:

Refer to the above Post.Attaching the Exception

**Comment by chrsmith** [comment]
Comment [#112](https://code.google.com/p/selenium/issues/detail?id=27#c112) originally posted by chrsmith on 2013-04-24T11:08:52.000Z:

You should try to update before posting an issue/adding to an existing one. v2.25 is rather old.

**Comment by chrsmith** [comment]
Comment [#113](https://code.google.com/p/selenium/issues/detail?id=27#c113) originally posted by chrsmith on 2013-12-04T06:06:06.000Z:


 WebElement actions =driver.findElement(By.id("actions"));
((JavascriptExecutor)driver).executeScript("arguments[0].click();" ,actions);


Alert alert = driver.switchTo().alert(); 
System.out.println("alert :::" + alert.getText()); 
alert.accept();

 the alert is populated. While debugging, I 
found that, the debug control is lost somewhere after I submit the 
form. However, in the application to be tested, the alert appears. 
Only after I click the alert manually, the debug control comes to 
Alert alert = driver.switchTo().alert(); statement. But since I have 
clicked on the alert already, the 
org.openqa.selenium.NoAlertPresentException is thrown.

**Comment by chrsmith** [comment]
Comment [#114](https://code.google.com/p/selenium/issues/detail?id=27#c114) originally posted by chrsmith on 2013-12-04T06:08:38.000Z:

pls anybody help me in fixing the above

**Comment by chrsmith** [comment]
Comment [#115](https://code.google.com/p/selenium/issues/detail?id=27#c115) originally posted by chrsmith on 2013-12-04T06:09:06.000Z:

pls anybody help me in fixing the above

**Comment by chrsmith** [comment]
Comment [#116](https://code.google.com/p/selenium/issues/detail?id=27#c116) originally posted by chrsmith on 2014-01-21T15:33:37.000Z:

The confirm box is not accepted when using ChromeDriver in the scenario bellow:

Have a web page (2.html) that opens a new web page (3.html) where a confirm box is shown onLoad event.

This works great with IE & Firefox but not with Chrome Driver - the confirm box stays unclosed and I have exception that is not presented.

Find attached archive with sample test pages and explanation of the scenario.

Regards

**Comment by chrsmith** [comment]
Comment [#117](https://code.google.com/p/selenium/issues/detail?id=27#c117) originally posted by chrsmith on 2014-05-14T07:00:00.000Z:


my requirement is 

->entering user id 
->entering password
->clicking on login button 
->if any one already login with same user id a popup message appears saying "already login with same user id"  
->clicking ok on popup
->entering password 
->click on "continue anyway" link

i able to accept popup message but how to write code in Java - webdrier when click on login button user should able to login but when popup message appears it should able to accept popup message and then click on "continue any way link 


code  

    driver.findElement(By.id("txtUserId")).click();
    driver.findElement(By.id("txtUserId")).sendKeys("prakashm");
    driver.findElement(By.id("txtPwd")).clear();
    driver.findElement(By.id("txtPwd")).sendKeys("cmc123");
    driver.findElement(By.id("btnLogin")).click();
    Alert alert = driver.switchTo().alert();
    alert.accept()
    driver.findElement(By.id("txtPwd")).clear();
    driver.findElement(By.id("txtPwd")).sendKeys("cmc123");
    driver.findElement(By.id("hlkReLogin")).click();

**Comment by chrsmith** [comment]
Comment [#118](https://code.google.com/p/selenium/issues/detail?id=27#c118) originally posted by chrsmith on 2014-06-24T12:52:23.000Z:

Hello,
I use Selenium 2.42.0 with Java and I need to use as WebDriver a headless browser like HtmlUnit or PhantomJS. It`s important that I can handle alert which i trigger on purpose for testing. I may close the Alert but I need to recognize it somehow in eclipse that there has been an Alert, like getting the Text of the alert or even an Exception that there is a alert present. 
 With this javascript i can only close the alert without recognition that there was an alert:
 if (!(driver instanceof HtmlUnitDriver)) {
       		System.out.println("test");
    		((JavascriptExecutor) driver).executeScript("window.alert = function(msg) {  return true; }");
    		((JavascriptExecutor) driver).executeScript("window.confirm = function(msg) { return true; }");
    		((JavascriptExecutor) driver).executeScript("window.prompt = function(msg) { return true; }");
    		}


Please is there any possibility just to recognize the alert in eclipse with java and HtmlUnit or PhantomJs with Selenium!!

**Comment by chrsmith** [comment]
Comment [#119](https://code.google.com/p/selenium/issues/detail?id=27#c119) originally posted by chrsmith on 2014-07-28T06:19:57.000Z:

Hi,

I am using selenium web driver ruby.I want to click on java script alert message (enter key) using ruby command.
Please provide me the solution.
